source/extensions/transport_sockets/alts/tsi_socket.cc

```diff
@@ -0,0 +1,214 @@
+/* Copyright 2018 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "extensions/transport_sockets/alts/tsi_socket.h"
+
+#include "common/common/assert.h"
+#include "common/common/enum_to_int.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Alts {
+
+TsiSocket::TsiSocket(HandshakerFactory handshaker_factory, HandshakeValidator handshake_validator)
+    : handshaker_factory_(handshaker_factory), handshake_validator_(handshake_validator),
+      raw_buffer_callbacks_(*this) {
+  raw_buffer_socket_.setTransportSocketCallbacks(raw_buffer_callbacks_);
+}
+
+TsiSocket::~TsiSocket() { ASSERT(!handshaker_); }
+
+void TsiSocket::setTransportSocketCallbacks(Envoy::Network::TransportSocketCallbacks& callbacks) {
+  callbacks_ = &callbacks;
+
+  handshaker_ = handshaker_factory_(callbacks.connection().dispatcher());
+  handshaker_->setHandshakerCallbacks(*this);
+}
+
+std::string TsiSocket::protocol() const { return ""; }
+
+Network::PostIoAction TsiSocket::doHandshake() {
+  ASSERT(!handshake_complete_);
+  ENVOY_CONN_LOG(debug, "TSI: doHandshake", callbacks_->connection());
+
+  if (!handshaker_next_calling_) {
+    doHandshakeNext();
+  }
+  return Network::PostIoAction::KeepOpen;
+}
+
+void TsiSocket::doHandshakeNext() {
+  ENVOY_CONN_LOG(debug, "TSI: doHandshake next: received: {}", callbacks_->connection(),
+                 raw_read_buffer_.length());
+  handshaker_next_calling_ = true;
+  Buffer::OwnedImpl handshaker_buffer;
+  handshaker_buffer.move(raw_read_buffer_);
+  handshaker_->next(handshaker_buffer);
+}
+
+Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result) {
+  ASSERT(next_result);
+
+  ENVOY_CONN_LOG(debug, "TSI: doHandshake next done: status: {} to_send: {}",
+                 callbacks_->connection(), next_result->status_, next_result->to_send_->length());
+
+  tsi_result status = next_result->status_;
+  tsi_handshaker_result* handshaker_result = next_result->result_.get();
+
+  if (status != TSI_INCOMPLETE_DATA && status != TSI_OK) {
+    ENVOY_CONN_LOG(debug, "TSI: Handshake failed: status: {}", callbacks_->connection(), status);
+    return Network::PostIoAction::Close;
+  }
+
+  if (next_result->to_send_->length() > 0) {
+    raw_write_buffer_.move(*next_result->to_send_);
```

Comment:
This may be too specific for this draft cl, but I have a concern that the data added to raw_writer_buffer might not have chance to be written if doHandshakeNextDone() is called asynchronously or it requires another write event to push the data. Please verify this case in your tests.